File: app/src/main/java/adapter/PersonPopularAdapter.java
package adapter;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;

import activity.PersonActivity;
import activity.PersonPopularActivity;
import br.com.icaro.filme.R;
import info.movito.themoviedbapi.TmdbPeople;
import info.movito.themoviedbapi.model.people.Person;
import utils.Constantes;
import utils.UtilsFilme;

/**
 * Created by icaro on 04/10/16.
 */
public class PersonPopularAdapter extends RecyclerView.Adapter<PersonPopularAdapter.HolderPersonPopular> {

    private Context context;
    private TmdbPeople.PersonResultsPage personResultsPage;

    public PersonPopularAdapter(PersonPopularActivity personPopular, TmdbPeople.PersonResultsPage personResultsPage) {
        context = personPopular;
        this.personResultsPage = personResultsPage;
    }

    @Override
    public HolderPersonPopular onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.adapter_person_popular, parent, false);
        return new HolderPersonPopular(view);
    }

    @Override
    public void onBindViewHolder(final HolderPersonPopular holder, final int position) {
        if (personResultsPage != null) {

            final Person person = personResultsPage.getResults().get(position);
            Picasso.with(context)
                    .load(UtilsFilme.getBaseUrlImagem(UtilsFilme.getTamanhoDaImagem(context, 2)) + person.getProfilePath())
                    .memoryPolicy(MemoryPolicy.NO_STORE, MemoryPolicy.NO_CACHE)
                    .networkPolicy(NetworkPolicy.NO_CACHE, NetworkPolicy.NO_STORE)
                    .into(holder.img_person, new Callback() {
                        @Override
                        public void onSuccess() {
                            holder.progressBar.setVisibility(View.GONE);
                        }

                        @Override
                        public void onError() {
                            holder.progressBar.setVisibility(View.GONE);
                            holder.text_person_name.setText(person.getName());
                            holder.text_person_name.setVisibility(View.VISIBLE);
                            holder.img_person.setImageResource(R.drawable.person);
                        }
                    });

            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Intent intent = new Intent(context, PersonActivity.class);
                    intent.putExtra(Constantes.PERSON_ID, person.getId());
                    intent.putExtra(Constantes.NOME_PERSON, person.getName());
                    context.startActivity(intent);
                }
            });
        }
    }

    @Override
    public int getItemCount() {

        return personResultsPage.getResults().size();
    }

    class HolderPersonPopular extends RecyclerView.ViewHolder {

        private ProgressBar progressBar;
        private ImageView img_person;
        private TextView text_person_name;

        HolderPersonPopular(View itemView) {
            super(itemView);
            progressBar = (ProgressBar) itemView.findViewById(R.id.progress);
            text_person_name = (TextView) itemView.findViewById(R.id.text_person_name);
            img_person = (ImageView) itemView.findViewById(R.id.img_popular_person);
        }
    }
}

File: app/src/main/java/pessoaspopulares/adapter/PersonPopularAdapter.kt
package pessoaspopulares.adapter

import android.support.v4.util.SparseArrayCompat
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import domain.PersonItem
import domain.ViewType
import utils.Constantes
import java.util.*

/**
 * Created by icaro on 04/10/16.
 */
class PersonPopularAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>(){


    private var personResultsPage = ArrayList<ViewType>()
    private var delegateAdapters = SparseArrayCompat<ViewTypeDelegateAdapter>()

    private val loadingItem = object : ViewType {
        override fun getViewType() = Constantes.BuscaConstants.LOADING
    }

    init {
        delegateAdapters.put(Constantes.BuscaConstants.LOADING, LoadingDelegateAdapter())
        delegateAdapters.put(Constantes.BuscaConstants.NEWS, PersonDelegateAdapter())
        personResultsPage = ArrayList()
        personResultsPage.add(loadingItem)
    }

    fun addPersonPopular(personResults: List<PersonItem?>?) {

        val initPosition = personResultsPage?.size!! - 1
        this.personResultsPage?.removeAt(initPosition)
        notifyItemRemoved(initPosition)

        // insert news and the loading at the end of the list
        for (person in personResults!!) {
            this.personResultsPage.add(person!!)
        }

        notifyItemRangeChanged(initPosition, this.personResultsPage?.size!! + 1 /* plus loading item */)
        personResultsPage.add(loadingItem)

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            delegateAdapters.get(viewType).onCreateViewHolder(parent)

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int) =
            delegateAdapters.get(getItemViewType(position)).onBindViewHolder(holder!!, personResultsPage[position], context = null)

    override fun getItemViewType(position: Int): Int = personResultsPage[position].getViewType()

    override fun getItemCount(): Int = personResultsPage.size
}



